src/katsdpmodels/primary_beam.py

```diff
@@ -234,7 +323,437 @@ def sample_grid(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noq
+        self._antenna = antenna
+        self._receiver = receiver
+        self._band = band
+
+    @property
+    def x(self) -> u.Quantity:
+        """x coordinates associated with the samples."""
+        return np.arange(self.samples.shape[-1]) * self.x_step + self.x_start
+
+    @property
+    def y(self) -> u.Quantity:
+        """y coordinates associated with the samples."""
+        return np.arange(self.samples.shape[-2]) * self.y_step + self.y_start
+
+    def spatial_resolution(self, frequency: u.Quantity) -> np.ndarray:
+        # Compute the Nyquist frequency, taking the maximum between x and y
+        x = self.x
+        y = self.y
+        scale = max(max(abs(x[0]), abs(x[-1])), max(abs(y[0]), y[-1]))
+        wavelength = frequency.to(u.m, equivalencies=u.spectral(), copy=False)
+        return (0.5 * wavelength / scale).to_value(u.dimensionless_unscaled)
+
+    def frequency_range(self) -> Tuple[u.Quantity, u.Quantity]:
+        return self.frequency[0], self.frequency[-1]
+
+    def frequency_resolution(self) -> u.Quantity:
+        return self._frequency_resolution
+
+    def inradius(self, frequency: u.Quantity) -> float:
+        wavelength = frequency.to(u.m, equivalencies=u.spectral(), copy=False)
+        # Assume that the model extends to the Nyquist limit
+        return 0.5 * float(wavelength / max(self.x_step, self.y_step))
+
+    def circumradius(self, frequency: u.Quantity) -> float:
+        wavelength = frequency.to(u.m, equivalencies=u.spectral(), copy=False)
+        return 0.5 * float(np.hypot(wavelength / self.x_step, wavelength / self.y_step))
+
+    @property
+    def is_circular(self) -> bool:
+        return False
+
+    @property
+    def is_unpolarized(self) -> bool:
+        return False
+
+    @property
+    def antenna(self) -> Optional[str]:
+        return self._antenna
+
+    @property
+    def receiver(self) -> Optional[str]:
+        return self._receiver
+
+    @property
+    def band(self) -> Optional[str]:
+        return self._band
+
+    def _prepare_samples(self, frequency: u.Quantity,
+                         frame: Union[AltAzFrame, RADecFrame],
+                         output_type: OutputType) -> np.ndarray:
+        """Interpolate aperture plane to selected frequencies, and optionally rotate."""
+        frequency_Hz = frequency.to_value(u.Hz).astype(np.float32, copy=False, casting='same_kind')
+        samples = self._interp_samples(frequency_Hz)
+        if output_type in {OutputType.JONES_XY, OutputType.MUELLER}:
+            if not isinstance(frame, RADecFrame):
+                raise ValueError('JONES_XY required a RADecFrame')
+            jones = frame.jones_hv_to_xy().astype(np.complex64, copy=False)
+            # Matrix multiply, but tensordot/matmul would require shuffling
+            # the axes around. TODO: see if it's faster to move into Numba
+            # inner loop.
+            samples = np.einsum('ij,...jkxy->...ikxy', jones, samples)
```

Comment:
I've not used the Einstein sum before: might be a good place for a comment?

Reply:
I've expanded the comment - let me know if it helps.